### DISCUSSION TITLE
How to send audio files using python-bale-bot?

### DISCUSSION THREAD

**Discussion by okaeiz** [kianahmadian/python-bale-bot#13] [Q&A] [answered]
Hi there!
I was wondering how I could send audio files using the library. 
For example,
when somebody has clicked a button and sent a callback, the answer of the callback will be an audio message.
**Do you have any samples?**

**Comment by kian-ahmadian** [comment, marked as answer]
Hey @okaeiz,
We haven't specifically written a method for it yet, but if you are coding in the `on_callback` event, you can use the callback object and go to the message and then access that chat and send an audio file in the chat with the file sending method. (`callback.message.chat`)

> Note: Due to the disturbances created in "Bale", it is not possible to upload through bots, but you can use the file_id of the sound to give it instead of the document.

```py
...
@bot.listen(bale.EventType.CALLBACK)
async def on_callback(callback: bale.CallbackQuery):
    await callback.message.chat.send_document(document = "Your file id", ...)
```

~~🔷**We will write a special method for it soon.**~~
**In the new version (`v2.4.4`), special methods were prepared**.

**Reply by okaeiz** [reply]
Thanks for your guide. I checked it out.
Here is a question. What is a file_id? How can I find the file_id of the file which I intend to send? Is that to say that we cannot send audio files?

**Reply by okaeiz** [reply]
This is what I have written:
```
"""keyboard handler"""
@client.listen(EventType.CALLBACK)
async def when_receive_callback(callback: CallbackQuery):
    if callback.data == "audio":
        audio = open("/path_to/audio_file.mp3","rb")
        await callback.message.reply_document(audio)
```
But when I run the code above and the users clicks on the button and the callback is sent, I recieve the following error:
```
error on_callback document param must be type of bytes, str or Document
```

**Reply by kian-ahmadian** [reply]
first, You should read the file.
```py
file = open("file path", "rb")
audio = file.read()
```

**Reply by okaeiz** [reply]
I highly appreciate your help. I really do.
I read the file - as you said - and changed the script like below:
```
"""keyboard handler"""
@client.listen(EventType.CALLBACK)
async def when_receive_callback(callback: CallbackQuery):
    if callback.data == "audio":
        audio_file = open("/path_to/audio_file.mp3","rb")
        audio = audio_file.read()
        await callback.message.reply_document(audio)
```
But now when the button is clicked on, the following error occurs:
```
error on_callback Can not serialize value type: <class 'int'>
 headers: {}
 value: 800646076
```

**Reply by kian-ahmadian** [reply]
This problem is caused by messenger, we are still following this problem.

**But in order to be able to do your work faster and not wait for our follow-up**, you can upload the file separately on Yes servers and then get its ID file using the bot. Then give it as a type string to the file id parameter.

**Reply by okaeiz** [reply]
Thanks.
So if the bot is restarted, do I have to send the files to it again?
Moreover, how can I obtain the file_id from the message? I found the `chat_id` and `message_id` only.

**Reply by kian-ahmadian** [reply]
I did not understand your first question

**But regarding the second question, when a message is sent by the user and you receive it in the on_message event, if you have sent a file in your message, the value of message.document is full, and it returns the document class. And you can reach file_id using it.**

Useful links about this:
[Document Class](https://python-bale-bot.readthedocs.io/en/master/api.html#document)
[Message Class](https://python-bale-bot.readthedocs.io/en/master/api.html#message)

**Reply by okaeiz** [reply]
I mean, if the server is rebooted, then do have to send the audio files to the bot once again?

**Reply by kian-ahmadian** [reply]
### **No, it is not necessary.**

The files you upload are in "Bale" servers, also the file IDs are the addresses of the files sent.

**Reply by okaeiz** [reply]
Could you please share the code snippet which handles the audio_file that is sent to the bot? I could not do it. I am sorry for asking too many questions; I am a Bale newbie.

**Reply by kian-ahmadian** [reply]
### **No problem, I am here to answer you.**

```python
...
@bot.listen(bale.EventType.MESSAGE)
async def on_callback(message: bale.Message):
    if message.document:
        print(message.document.file_id)
```

**Reply by okaeiz** [reply]
Thanks for that, it works fine right now.
Using this method I could send photos, audio files and video files.
There's an issue with sending video files and GIFs, though. However, I'll create a new discussion for that.
